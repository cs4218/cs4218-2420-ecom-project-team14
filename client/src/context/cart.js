import axios from "axios";
import { createContext, useContext, useEffect, useReducer } from "react";
import toast from "react-hot-toast";
import { useAuth } from "./auth";

const CartContext = createContext();

const cartReducer = (state, action) => {
  switch (action.type) {
    case "ADD_TO_CART": {
      const { slug, price, productId } = action.payload;
      toast.success("Add to Cart Successfully");
      return {
        ...state,
        [slug]: state[slug]
          ? { quantity: state[slug].quantity + 1, price, productId}
          : { quantity: 1, price, productId},
      };
    }
    case "REMOVE_FROM_CART": {
      const newState = { ...state };
      delete newState[action.payload.slug];
      toast.success("Remove from Cart Successfully");
      return newState;
    }
    case "UPDATE_QUANTITY": {
      const { slug, quantity, price, productId } = action.payload;
      if (quantity <= 0) {
        const newState = { ...state };
        delete newState[slug];
        return newState;
      }
      toast.success("Update Cart Quantity Successfully");
      return {
        ...state,
        [slug]: { ...state[slug], quantity, price, productId },
      };
    }
    case "CLEAR_CART":
      toast.success("Cart Cleared Successfully");
      return {};
    default:
      return state;
  }
};

export const CartProvider = ({ children }) => {
  const auth = useAuth();

  const user = auth.user ?? "guest";

  const [cart, dispatch] = useReducer(cartReducer, {}, () => {
    const localData = localStorage.getItem(`cart-${user}`);
    return localData ? JSON.parse(localData) : {};
  });

  useEffect(() => {
    localStorage.setItem(`cart-${user}`, JSON.stringify(cart));
  }, [cart]);

  const addToCart = async (slug) => {
    const { data } = await axios.get(`/api/v1/product/get-product/${slug}`);
    if (!data.product) {
      toast.error("Item does not exist");
      return;
    }

    const inventory = data.product.quantity;

    if (cart[slug] && cart[slug].quantity + 1 > inventory) {
      toast.error("Error added to cart: Not enough inventory");
      return;
    }

    if (!data.product.price) {
      toast.error("Error added to cart: Price of product not available");
      return;
    }

    dispatch({ type: "ADD_TO_CART", payload: { slug, price: data.product.price, productId: data.product._id } });
  };

  const removeFromCart = (slug) => {
    dispatch({ type: "REMOVE_FROM_CART", payload: { slug } });
  };

  const updateQuantity = async (slug, quantity) => {
    const { data } = await axios.get(`/api/v1/product/get-product/${slug}`);
    if (!data.product) {
      toast.error("Item does not exist");
      return;
    }

    const inventory = data.product.quantity;

    if (cart[slug] && quantity > inventory) {
      toast.error("Error updating quantity: Not enough inventory");
      return;
    }

    if (!data.product.price) {
      toast.error("Error added to cart: Price of product not available");
      return;
    }

    dispatch({ type: "UPDATE_QUANTITY", payload: { slug, quantity, price: data.product.price, productId: data.product._id } });
  };

  const clearCart = () => {
    dispatch({ type: "CLEAR_CART" });
  };

  return (
    <CartContext.Provider
      value={{
        cart,
        addToCart,
        removeFromCart,
        updateQuantity,
        clearCart,
      }}
    >
      {children}
    </CartContext.Provider>
  );
};

export const useCart = () => {
  return useContext(CartContext);
};
