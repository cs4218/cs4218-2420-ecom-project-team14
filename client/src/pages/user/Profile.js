import React, { useState, useEffect } from "react";
import UserMenu from "../../components/UserMenu";
import Layout from "./../../components/Layout";
import { useAuth } from "../../context/auth";
import toast from "react-hot-toast";
import axios from "axios";
const Profile = () => {
  //context
  const [auth, setAuth] = useAuth();
  //state
  const [name, setName] = useState("");
  const [email, setEmail] = useState("");
  const [password, setPassword] = useState("");
  const [phone, setPhone] = useState("");
  const [address, setAddress] = useState("");
  const [errors, setErrors] = useState({});

  //get user data
  useEffect(() => {
    const { email, name, phone, address } = auth?.user;
    setName(name);
    setPhone(phone);
    setEmail(email);
    setAddress(address);
  }, [auth?.user]);

  const validateData = () => {
    let errors = {};

    // Check that the length of the name is at most 150 characters long
    if (name.length > 150) {
        errors.name = "The length of the name can only be up to 150 characters";
    }

    // Check that the phone number is 8 digits long and start with 6,8 or 9
    const phoneRegex = /^[689]\d{7}$/;
    if (!phone.match(phoneRegex)) {
      errors.phone = "The phone number must start with 6,8 or 9 and be 8 digits long";
    }

    // Check that new password is at least of length 6 if not empty
    if (password && password.length < 6) {
        errors.password = "The length of the new password needs to be at least of length 6";
    }

    // Check that the address is at most 150 characters long
    if (address.length > 150) {
      errors.address = "The address can only be up to 150 characters long";
    }

    return errors
};

  // form function
  const handleSubmit = async (e) => {
    e.preventDefault();
    setErrors({});
    const errorList = validateData();

    if (Object.entries(errorList).length > 0) {
        setErrors(errorList);
        toast.error("There are invalid fields in the update profile form")
        return;
    }

    const trimmedName = name.trim();
    const trimmedPassword = password.trim();
    const trimmedPhone = phone.trim();
    const trimmedAddress = address.trim();

    try {
      const { data } = await axios.put("/api/v1/auth/profile", {
        name: trimmedName,
        password: trimmedPassword,
        phone: trimmedPhone,
        address: trimmedAddress,
      });
      if (data?.error) {
        toast.error(data?.error);
      } else {
        setAuth({ ...auth, user: data?.updatedUser });
        let ls = localStorage.getItem("auth");
        ls = JSON.parse(ls);
        ls.user = data.updatedUser;
        localStorage.setItem("auth", JSON.stringify(ls));
        toast.success("Profile Updated Successfully");
      }
    } catch (error) {
      if (error.response && error.response.status === 400) {
        toast.error(error.response.data?.message ?? "Error encountered when updating profile");
      } else {
        console.log(error);
        toast.error("Something went wrong");
      }
    }
  };
  return (
    <Layout title={"Your Profile"}>
      <div className="container-fluid m-3 p-3">
        <div className="row">
          <div className="col-md-3">
            <UserMenu />
          </div>
          <div className="col-md-9">
            <div className="form-container ">
              <form onSubmit={handleSubmit}>
                <h4 className="title">USER PROFILE</h4>
                <div className="mb-3">
                  <input
                    type="text"
                    value={name}
                    onChange={(e) => setName(e.target.value)}
                    className="form-control"
                    id="exampleInputEmail1"
                    placeholder="Enter Your Name (Required)"
                    required
                    autoFocus
                  />
                  {errors.name && (
                  <p style={{ color: "red" }}>
                      {errors.name}
                  </p>)}
                </div>
                <div className="mb-3">
                  <input
                    type="email"
                    value={email}
                    onChange={(e) => setEmail(e.target.value)}
                    className="form-control"
                    id="exampleInputEmail1"
                    placeholder="Enter Your Email"
                    disabled
                  />
                </div>
                <div className="mb-3">
                  <input
                    type="password"
                    value={password}
                    onChange={(e) => setPassword(e.target.value)}
                    className="form-control"
                    id="exampleInputPassword1"
                    placeholder="Enter Your Password"
                  />
                  {errors.password && (
                  <p style={{ color: "red" }}>
                      {errors.password}
                  </p>)}
                </div>
                <div className="mb-3">
                  <input
                    type="text"
                    value={phone}
                    onChange={(e) => setPhone(e.target.value)}
                    className="form-control"
                    id="exampleInputEmail1"
                    required
                    placeholder="Enter Your Phone (Required)"
                  />
                  {errors.phone && (
                  <p style={{ color: "red" }}>
                      {errors.phone}
                  </p>)}
                </div>
                <div className="mb-3">
                  <input
                    type="text"
                    value={address}
                    onChange={(e) => setAddress(e.target.value)}
                    className="form-control"
                    id="exampleInputEmail1"
                    required
                    placeholder="Enter Your Address (Required)"
                  />
                  {errors.address && (
                  <p style={{ color: "red" }}>
                      {errors.address}
                  </p>)}
                </div>

                <button type="submit" className="btn btn-primary">
                  UPDATE
                </button>
              </form>
            </div>
          </div>
        </div>
      </div>
    </Layout>
  );
};

export default Profile;