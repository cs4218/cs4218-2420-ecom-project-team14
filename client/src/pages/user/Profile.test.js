import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { act } from "react-dom/test-utils";
import "@testing-library/jest-dom";
import Profile from "./Profile";
import { useAuth } from "../../context/auth";
import axios from "axios";
import toast from "react-hot-toast";

// Mock dependencies
jest.mock("../../components/Layout", () => ({ children, title }) => (
  <div data-testid="mock-layout" data-title={title}>
    {children}
  </div>
));

jest.mock("../../components/UserMenu", () => () => (
  <div data-testid="mock-user-menu">User Menu</div>
));

jest.mock("../../context/auth", () => ({
  useAuth: jest.fn(),
}));

jest.mock("axios");
jest.mock("react-hot-toast");

// Mock localStorage
Object.defineProperty(window, "localStorage", {
  value: {
    getItem: jest.fn(() => JSON.stringify({ user: {}, token: "123" })),
    setItem: jest.fn(),
  },
  writable: true,
});

describe("Profile Component", () => {
  const mockUser = {
    name: "John Doe",
    email: "john@example.com",
    phone: "1234567890",
    address: "123 Test Street",
  };

  const mockSetAuth = jest.fn();

  beforeEach(() => {
    jest.clearAllMocks();
    useAuth.mockReturnValue([{ user: mockUser }, mockSetAuth]);
  });

  // Basic Rendering Tests
  describe("Component Rendering", () => {
    it("renders Profile component without crashing", () => {
      render(<Profile />);
      expect(screen.getByTestId("mock-layout")).toBeInTheDocument();
      expect(screen.getByTestId("mock-user-menu")).toBeInTheDocument();
    });

    it("renders with correct layout title", () => {
      render(<Profile />);
      expect(screen.getByTestId("mock-layout")).toHaveAttribute(
        "data-title",
        "Your Profile"
      );
    });

    it("displays the profile form with correct title", () => {
      render(<Profile />);
      expect(screen.getByText("USER PROFILE")).toBeInTheDocument();
    });
  });

  // User Data Loading Tests
  describe("User Data Loading", () => {
    it("loads user data from auth context", () => {
      render(<Profile />);
      
      // Check if form fields are pre-filled with user data
      expect(screen.getByDisplayValue(mockUser.name)).toBeInTheDocument();
      expect(screen.getByDisplayValue(mockUser.email)).toBeInTheDocument();
      expect(screen.getByDisplayValue(mockUser.phone)).toBeInTheDocument();
      expect(screen.getByDisplayValue(mockUser.address)).toBeInTheDocument();
    });

    it("disables the email field", () => {
      render(<Profile />);
      const emailField = screen.getByDisplayValue(mockUser.email);
      expect(emailField).toBeDisabled();
    });
  });

  // Form Interaction Tests
  describe("Form Interactions", () => {
    it("allows updating user name", () => {
      render(<Profile />);
      const nameInput = screen.getByDisplayValue(mockUser.name);
      fireEvent.change(nameInput, { target: { value: "Jane Doe" } });
      expect(nameInput.value).toBe("Jane Doe");
    });

    it("allows updating user password", () => {
      render(<Profile />);
      const passwordInput = screen.getByPlaceholderText("Enter Your Password");
      fireEvent.change(passwordInput, { target: { value: "newpassword123" } });
      expect(passwordInput.value).toBe("newpassword123");
    });

    it("allows updating user phone number", () => {
      render(<Profile />);
      const phoneInput = screen.getByDisplayValue(mockUser.phone);
      fireEvent.change(phoneInput, { target: { value: "9876543210" } });
      expect(phoneInput.value).toBe("9876543210");
    });

    it("allows updating user address", () => {
      render(<Profile />);
      const addressInput = screen.getByDisplayValue(mockUser.address);
      fireEvent.change(addressInput, { target: { value: "456 New Street" } });
      expect(addressInput.value).toBe("456 New Street");
    });
  });

  // Form Submission Tests
  describe("Form Submission", () => {
    it("submits updated profile successfully", async () => {
      const updatedUser = {
        ...mockUser,
        name: "Jane Doe",
        phone: "98765432",
        address: "456 New Street",
      };

      axios.put.mockResolvedValueOnce({
        data: {
          updatedUser,
        },
      });

      render(<Profile />);

      // Update form fields
      fireEvent.change(screen.getByDisplayValue(mockUser.name), {
        target: { value: updatedUser.name },
      });
      fireEvent.change(screen.getByDisplayValue(mockUser.phone), {
        target: { value: updatedUser.phone },
      });
      fireEvent.change(screen.getByDisplayValue(mockUser.address), {
        target: { value: updatedUser.address },
      });
      
      // Submit the form
      fireEvent.click(screen.getByText("UPDATE"));

      await waitFor(() => {
        expect(axios.put).toHaveBeenCalledWith("/api/v1/auth/profile", {
          name: updatedUser.name,
          password: "",
          phone: updatedUser.phone,
          address: updatedUser.address,
        });
        
        expect(mockSetAuth).toHaveBeenCalledWith({
          user: updatedUser,
        });
        
        expect(localStorage.setItem).toHaveBeenCalled();
        expect(toast.success).toHaveBeenCalledWith("Profile Updated Successfully");
      });
    });

    it("handles API error with error response", async () => {
      axios.put.mockRejectedValueOnce({
        response: {
          status: 400,
          data: {
            message: "Passsword is required and 6 character long",
          },
        },
      });
      const updatedUser = {
        name: "Captain Underpants",
        phone: "98765432",
        password: " 12345",
        address: "456 New Street",
      };

      render(<Profile />);
      
      // Update form fields
      fireEvent.change(screen.getByDisplayValue(mockUser.name), {
        target: { value: updatedUser.name },
      });
      fireEvent.change(screen.getByDisplayValue(mockUser.phone), {
        target: { value: updatedUser.phone },
      });
      fireEvent.change(screen.getByDisplayValue(mockUser.address), {
        target: { value: updatedUser.address },
      });
      fireEvent.change(screen.getByDisplayValue(""), {
        target: { value: updatedUser.password },
      });
      
      fireEvent.click(screen.getByText("UPDATE"));

      await waitFor(() => {
        expect(toast.error).toHaveBeenCalledWith("Passsword is required and 6 character long");
      });
    });

    it("handles generic API error", async () => {
      axios.put.mockRejectedValueOnce(new Error("Network error"));

      render(<Profile />);

      const updatedUser = {
        name: "Captain Underpants",
        phone: "98765432",
        address: "456 New Street",
      };

      // Update form fields
      fireEvent.change(screen.getByDisplayValue(mockUser.name), {
        target: { value: updatedUser.name },
      });
      fireEvent.change(screen.getByDisplayValue(mockUser.phone), {
        target: { value: updatedUser.phone },
      });
      fireEvent.change(screen.getByDisplayValue(mockUser.address), {
        target: { value: updatedUser.address },
      });
      
      fireEvent.click(screen.getByText("UPDATE"));

      await waitFor(() => {
        expect(toast.error).toHaveBeenCalledWith("Something went wrong");
      });
    });
  });

  // Structure Tests
  describe("Component Structure", () => {
    it("contains main container with bootstrap classes", () => {
      const { container } = render(<Profile />);
      const mainContainer = container.querySelector(".container-fluid");
      expect(mainContainer).toBeInTheDocument();
      expect(mainContainer).toHaveClass("m-3", "p-3");
    });

    it("contains two columns with correct classes", () => {
      const { container } = render(<Profile />);
      const col1 = container.querySelector(".col-md-3");
      const col2 = container.querySelector(".col-md-9");
      expect(col1).toBeInTheDocument();
      expect(col2).toBeInTheDocument();
    });
  });

  // Snapshot Test
  describe("Snapshot", () => {
    it("matches snapshot", () => {
      const { container } = render(<Profile />);
      expect(container).toMatchSnapshot();
    });
  });
});