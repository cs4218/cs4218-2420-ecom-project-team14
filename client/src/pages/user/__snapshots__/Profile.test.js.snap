// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile Component Snapshot matches snapshot 1`] = `
<div>
  <div
    data-testid="mock-layout"
    data-title="Your Profile"
  >
    <div
      class="container-fluid m-3 p-3"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          <div
            data-testid="mock-user-menu"
          >
            User Menu
          </div>
        </div>
        <div
          class="col-md-9"
        >
          <div
            class="form-container "
          >
            <form>
              <h4
                class="title"
              >
                USER PROFILE
              </h4>
              <div
                class="mb-3"
              >
                <input
                  class="form-control"
                  id="exampleInputEmail1"
                  placeholder="Enter Your Name"
                  required=""
                  type="text"
                  value="John Doe"
                />
              </div>
              <div
                class="mb-3"
              >
                <input
                  class="form-control"
                  disabled=""
                  id="exampleInputEmail1"
                  placeholder="Enter Your Email "
                  type="email"
                  value="john@example.com"
                />
              </div>
              <div
                class="mb-3"
              >
                <input
                  class="form-control"
                  id="exampleInputPassword1"
                  placeholder="Enter Your Password"
                  type="password"
                  value=""
                />
              </div>
              <div
                class="mb-3"
              >
                <input
                  class="form-control"
                  id="exampleInputEmail1"
                  placeholder="Enter Your Phone"
                  required=""
                  type="text"
                  value="1234567890"
                />
              </div>
              <div
                class="mb-3"
              >
                <input
                  class="form-control"
                  id="exampleInputEmail1"
                  placeholder="Enter Your Address"
                  required=""
                  type="text"
                  value="123 Test Street"
                />
              </div>
              <button
                class="btn btn-primary"
                type="submit"
              >
                UPDATE
              </button>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
